src/background/providers/hcaptcha.ts

```diff
@@ -24,19 +81,179 @@ export class HcaptchaProvider implements Provider {
+                'cf-chl-bypass': this.getID().toString(),
+            };
+
+            const response = await axios.post<string, { data: string }>(url, requestBody, {
+                headers,
+                responseType: 'text',
+            });
+
+            const { signatures } = qs.parse(response.data);
+            if (signatures === undefined) {
+                throw new Error('There is no signatures parameter in the issuance response.');
+            }
+            if (typeof signatures !== 'string') {
+                throw new Error(
+                    'The signatures parameter in the issuance response is not a string.',
+                );
+            }
+
+            const data: SignaturesParam = JSON.parse(atob(signatures));
+            const returned = voprf.getCurvePoints(data.sigs);
+            const commitment = await this.getCommitment(data.version);
+            const prng = data.prng || 'shake';
+
+            const result = voprf.verifyProof(
+                data.proof,
+                newTokens.map((token) => token.toLegacy()),
+                returned,
+                commitment,
+                prng,
+            );
+
+            if (!result) {
+                throw new Error('DLEQ proof is invalid.');
+            }
+
+            newTokens.forEach((token, index) => {
+                token.setSignedPoint(returned.points[index as number]);
+            });
+            const oldTokens = this.getStoredTokens();
+            this.setStoredTokens(oldTokens.concat(newTokens));
+            this.forceUpdateIcon();
+        })();
+    }
+
     handleBeforeRequest(
         _details: chrome.webRequest.WebRequestBodyDetails,
     ): chrome.webRequest.BlockingResponse | void {
         return;
     }
+
     handleBeforeSendHeaders(
-        _details: chrome.webRequest.WebRequestHeadersDetails,
+        details: chrome.webRequest.WebRequestHeadersDetails,
     ): chrome.webRequest.BlockingResponse | void {
+        const url = new URL(details.url);
+        const urlType = this.handleUrl(url);
+
+        if (urlType.isIssuerUrl && details.method.toLowerCase() === 'post') {
+            // Erase any previous attempt
+            this.issueInfo = null;
+
+            // Do not store infinite tokens
+            const tokens = this.getStoredTokens();
+            if (tokens.length + NUM_OF_REQ_TOKENS > MAX_NUM_OF_TOKENS) return;
+
+            this.issueInfo = {
+                newUrl: details.url,
+                tabId: details.tabId,
+            };
+            return;
+        } else if (urlType.isSpendUrl) {
+            // Get one token
+            const tokens = this.getStoredTokens();
+            const oneToken = tokens.shift();
+            if (oneToken === undefined) return;
+            this.setStoredTokens(tokens);
+            this.forceUpdateIcon();
+
+            // Spend logic here
+            const httpPath = details.method + ' ' + url.pathname;
```

Comment:
I feel this is a place where string interpolation is more readable.
```suggestion
            const httpPath = `${details.method} ${url.pathname}`;
```